disttask: make task executor `onError` print error's stack.

LindaSummer opened the pull request:
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: close #56014

Problem Summary:

### What changed and how does it work?

Error log in `onError` function's stack is always in the same function.

But the real stack should be the one invoking `onError`.

### Check List

Tests <!-- At least one of them must be included. -->

- [x] Unit test


### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

LindaSummer commented:
Hi @lance6716 ,

Please take a look. 😊

Thanks very much.

Best Regards,
Edward

lance6716 commented:
/ok-to-test

And please post the test result. Or if you can, add an unit test for it. For example, check the output of logger and verify it contents the function that throws the error

LindaSummer commented:
> /ok-to-test
> 
> And please post the test result. Or if you can, add an unit test for it. For example, check the output of logger and verify it contents the function that throws the error

Thanks very much for your suggestions!

I will try to add a test case for it. 😊

Best Regards,
Edward

LindaSummer commented:
Hi @lance6716 ,

I have corrected my mistake and add unit tests. 😊

Please take a look.

Best Regards,
Edward

lance6716 commented on pkg/disttask/framework/taskexecutor/task_executor_test.go:
```diff
@@ -546,3 +549,33 @@ func TestInject(t *testing.T) {
+	stack := errLog.ContextMap()["stack"]
+	require.IsType(t, "", stack)
+	stackStr := stack.(string)
+	require.Truef(t, strings.HasPrefix(stackStr,
```
Hi, the problem is if the error that passed to `onError` lost its trace. You can reproduce it like

```golang
func throwError() error {
	return errors.New("mock error")
}

func callOnError(taskExecutor *BaseTaskExecutor) {
	taskExecutor.onError(errors.Trace(throwError()))
}
```

the first line on stack should be `throwError` function

LindaSummer replied:
Hi @lance6716 ,

Please correct me if I misunderstand.

In my knowledge, the order of popping stack frame should be `throwError` -> `errors.Trace` -> `taskExecutor.onError`.

So, when the `errors.Trace` is invoked, the previous stack info has been popped and can't be retrieved again.

The only way is explicitly adding `errors.Trace` before popping `throwError`.

I have added the `errors.Trace` before all `onError`.

Best Regards,
Edward

LindaSummer replied:
Got it!

Thanks very much for your help and suggestions! 😊

I will check the `DisableErrorVerbose` option and the stack inside the error itself.

Best Regards,
Edward

lance6716 replied:
Just to remind that the user may turn off `DisableErrorVerbose` in config. We should find another way to print error stack.

LindaSummer replied:
Hi @lance6716 ,

I think the stable way to print the top stack should be the error field itself.

Since the stack field works now, I will try to embed the top stack info inside the error message itself.

Best Regards,
Edward

lance6716 reviewed (commented):
Please run `make bazel_prepare` to update `.bazel` files, or just apply the following git diff (run `git apply -` and paste following contents):
diff --git a/pkg/disttask/framework/taskexecutor/BUILD.bazel b/pkg/disttask/framework/taskexecutor/BUILD.bazel
index e65cc685e0..79a2483481 100644
--- a/pkg/disttask/framework/taskexecutor/BUILD.bazel
+++ b/pkg/disttask/framework/taskexecutor/BUILD.bazel
@@ -51,7 +51,7 @@ go_test(
     ],
     embed = [":taskexecutor"],
     flaky = True,
-    shard_count = 16,
+    shard_count = 17,
     deps = [
         "//pkg/disttask/framework/mock",
         "//pkg/disttask/framework/mock/execute",
@@ -74,5 +74,7 @@ go_test(
         "@org_golang_google_grpc//status",
         "@org_uber_go_goleak//:goleak",
         "@org_uber_go_mock//gomock",
+        "@org_uber_go_zap//:zap",
+        "@org_uber_go_zap//zaptest/observer",
     ],
 )

LindaSummer commented:
Thanks very much for your warm suggestion and help!

I will update this PR later today.😄

Best Regards,
Edward

LindaSummer commented:
Hi @lance6716 ,

Thanks very much for your bazel patch!😊

I have updated related code and added stacktrace explicitly for all `onError` invokers.

Best Regards,
Edward

LindaSummer commented:
Hi @lance6716 ，

Thanks for your patience and warm help! 😊

I have added an `error stack` field for stack info inside error and skipped it for error without stack info.

Best Regards,
Edward

LindaSummer commented:
/retest

lance6716 commented on pkg/disttask/framework/taskexecutor/task_executor_test.go:
```diff
@@ -546,3 +549,72 @@ func TestInject(t *testing.T) {
+		stackStr := stack.(string)
+		require.NotContains(t, contextMap, "error stack")
+		require.Truef(t, strings.HasPrefix(stackStr,
+			"github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor.callOnError"),
```
check `		require.NotContains(t, contextMap, "error stack")
`

lance6716 commented on pkg/disttask/framework/taskexecutor/task_executor.go:
```diff
@@ -532,8 +533,13 @@ func (e *BaseTaskExecutor) onError(err error) {
-	e.logger.Error("onError", zap.Error(err), zap.Stack("stack"))
+
+	if errors.HasStack(err) {
+		e.logger.Error("onError", zap.Error(err), zap.StackSkip("stack", 1), zap.String("error stack", fmt.Sprintf("%+v", err)))
```
```suggestion
		e.logger.Error("onError", zap.Error(err), zap.Stack("stack"), zap.String("error stack", fmt.Sprintf("%+v", err)))
```

no need to change this. The top of stack can be `onError`. The issue is not related to "stack" field but "error stack" field. Also change below branch and remove the test for "stack"

lance6716 reviewed (commented):
rest lgtm

LindaSummer commented:
> rest lgtm

Thanks very much for your patience and help! 😊

I'll update it later today.

Best Regards,
Edward

LindaSummer commented:
/retest

LindaSummer commented:
Hi @lance6716 ,

Thanks for your review and suggestions! 😊

I have updated the related code, please help take a look. 

Best Regards,
Edward

lance6716 commented on pkg/disttask/framework/taskexecutor/task_executor.go:
```diff
@@ -345,7 +346,7 @@ func (e *BaseTaskExecutor) runStep(resource *proto.StepResource) (resErr error)
 				e.logger.Info("subtask in running state and is not idempotent, fail it",
 					zap.Int64("subtask-id", subtask.ID))
-				e.onError(ErrNonIdempotentSubtask)
+				e.onError(errors.Trace(ErrNonIdempotentSubtask))
```
```suggestion
				e.onError(ErrNonIdempotentSubtask)
```

lance6716 commented on pkg/disttask/framework/taskexecutor/task_executor.go:
```diff
@@ -431,7 +432,7 @@ func (e *BaseTaskExecutor) runSubtask(ctx context.Context, stepExecutor execute.
 	failpoint.Inject("MockExecutorRunErr", func(val failpoint.Value) {
 		if val.(bool) {
-			e.onError(errors.New("MockExecutorRunErr"))
+			e.onError(errors.Trace(errors.New("MockExecutorRunErr")))
```
```suggestion
			e.onError(errors.New("MockExecutorRunErr"))
```

lance6716 commented on pkg/disttask/framework/taskexecutor/task_executor.go:
```diff
@@ -532,8 +533,13 @@ func (e *BaseTaskExecutor) onError(err error) {
+	if errors.HasStack(err) {
+		e.logger.Error("onError", zap.Error(err), zap.Stack("stack"), zap.String("error stack", fmt.Sprintf("%+v", err)))
+	} else {
+		err = errors.Trace(err)
```
```suggestion
```

lance6716 commented on pkg/disttask/framework/taskexecutor/task_executor.go:
```diff
@@ -532,8 +533,13 @@ func (e *BaseTaskExecutor) onError(err error) {
+		e.logger.Error("onError", zap.Error(err), zap.Stack("stack"), zap.String("error stack", fmt.Sprintf("%+v", err)))
+	} else {
+		err = errors.Trace(err)
+		e.logger.Error("onError", zap.Error(err), zap.StackSkip("stack", 1))
```
```suggestion
		e.logger.Error("onError", zap.Error(err), zap.Stack("stack"))
```

lance6716 commented:
/cc @GMHDBJD

LindaSummer commented:
/retest

LindaSummer commented:
Hi @GMHDBJD, 

Sorry to bother you. 😊

Could help take a look?

Best Regards,
Edward

LindaSummer commented:
/retest

LindaSummer commented:
/restest

LindaSummer commented:
/retest

LindaSummer commented:
/retest

Benjamin2037 reviewed (approved):
LGTM

LindaSummer commented:
/assign

D3Hunter commented on pkg/disttask/framework/taskexecutor/task_executor.go:
```diff
@@ -532,8 +533,12 @@ func (e *BaseTaskExecutor) onError(err error) {
-	err = errors.Trace(err)
-	e.logger.Error("onError", zap.Error(err), zap.Stack("stack"))
+
+	if errors.HasStack(err) {
```
why check it?

lance6716 replied:
we don't need to print the plain error stack. There's already zap.Error(err)

D3Hunter commented:
/unhold
